MoltenVK/MoltenVK/Commands/MVKCommandBuffer.mm

```diff
@@ -523,43 +532,42 @@
+	// and render areas, making this internal tile memory allocation extremely wasteful, occasionally
+	// to the point of triggering OOM crashes. In this case, if we have a viewport, use it as a more
+	// accurate representation of what the tiled render area will actually be.
 	bool hasAttachments = _attachments.size() > 0;
-	if (hasAttachments) {
-		VkExtent2D fbExtent = getFramebufferExtent();
-		mtlRPDesc.renderTargetWidthMVK = max(min(_renderArea.offset.x + _renderArea.extent.width, fbExtent.width), 1u);
-		mtlRPDesc.renderTargetHeightMVK = max(min(_renderArea.offset.y + _renderArea.extent.height, fbExtent.height), 1u);
-	}
-    if (_canUseLayeredRendering) {
-        uint32_t renderTargetArrayLength;
+	VkRect2D rendArea = clipToRenderArea(!hasAttachments && _viewportState.hasViewport()
+										 ? _viewportState.getUnionVkRect()
+										 : mvkMakeVkRect({0,0}, getFramebufferExtent()));
+	mtlRPDesc.renderTargetWidthMVK = max(rendArea.offset.x + rendArea.extent.width, 1u);
+	mtlRPDesc.renderTargetHeightMVK = max(rendArea.offset.y + rendArea.extent.height, 1u);
+
+	if (_canUseLayeredRendering) {
         bool found3D = false, found2D = false;
         for (uint32_t i = 0; i < 8; i++) {
             id<MTLTexture> mtlTex = mtlRPDesc.colorAttachments[i].texture;
-            if (mtlTex == nil) { continue; }
-            switch (mtlTex.textureType) {
-                case MTLTextureType3D:
-                    found3D = true;
-                default:
-                    found2D = true;
-            }
+			if (mtlTex) {
+				switch (mtlTex.textureType) {
+					case MTLTextureType3D:
+						found3D = true;
+					default:
+						found2D = true;
+				}
+			}
         }
 
-        if (getSubpass()->isMultiview()) {
-            // In the case of a multiview pass, the framebuffer layer count will be one.
-            // We need to use the view count for this multiview pass.
-			renderTargetArrayLength = getSubpass()->getViewCountInMetalPass(_multiviewPassIndex);
-        } else {
-			renderTargetArrayLength = getFramebufferLayerCount();
-        }
-        // Metal does not allow layered render passes where some RTs are 3D and others are 2D.
-        if (hasAttachments && (!(found3D && found2D) || renderTargetArrayLength > 1)) {
-            mtlRPDesc.renderTargetArrayLengthMVK = renderTargetArrayLength;
-        }
+		// As above, to avoid overallocation, do not use layering if there are no attachments.
+		// And Metal does not support layered render passes with a mix of 2D & 3D targets.
+		// For a multiview pass, the framebuffer layer count will be 1, so we need to use
+		// the view count for this multiview pass. Otherwise, use the framebuffer layer count.
+		if (hasAttachments && !(found3D && found2D)) {
+			mtlRPDesc.renderTargetArrayLengthMVK = (getSubpass()->isMultiview()
+													? getSubpass()->getViewCountInMetalPass(_multiviewPassIndex)
+													: getFramebufferLayerCount());
+		}
```

Comment:
Fair enough. 

What kind of test would we use to determine that?

Before I started messing with all this, originally, this test was:

        // Metal does not allow layered render passes where some RTs are 3D and others are 2D.
        if ( !(found3D && found2D) || renderTargetArrayLength > 1 ) {

This test seemed odd to me, because the the comment indicates we can't handle a mix of 2D & 3D attachments, but then we override that if the attachments have more than one layer. I don't understand the connection between the two conditions. They seem to be orthogonal. That's why I removed the layer override, to clarify that we don't set render layer count if we don't have any attachments, or if the attachment mix is unsupported.

At the same time, in issue #1650 which catalyzed this fix attempt, when not using attachments, FFXIV was setting the FB layer count to 2048, which triggers a large unnecessary tile memory allocation if we accept it blindly.

One of the DXVK hacks around this overallocation issue, was to say if we have no attachments, then set the layer count to 1, and I was following that thinking with `hasAttachments`.

With all that in mind, how do we figure out whether the app _really_ does want the extra layers?